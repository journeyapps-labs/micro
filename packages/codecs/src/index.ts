export * from './parsers';
export * from './codecs';
