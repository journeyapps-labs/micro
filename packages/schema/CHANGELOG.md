# @journeyapps-labs/micro-schema

## 1.0.1

### Patch Changes

- 1f54327: Bump all deps and fix types due to TS5.9
- Updated dependencies [1f54327]
  - @journeyapps-labs/micro-codecs@1.0.1
  - @journeyapps-labs/micro-errors@1.0.1

## 1.0.0

### Major Changes

- 3b6c502: Initial publish

### Patch Changes

- Updated dependencies [3b6c502]
  - @journeyapps-labs/micro-codecs@1.0.0
  - @journeyapps-labs/micro-errors@1.0.0
