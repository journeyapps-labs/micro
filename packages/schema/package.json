{
  "name": "@journeyapps-labs/micro-schema",
  "main": "./dist/index",
  "typings": "./dist/index",
  "version": "1.0.0",
  "publishConfig": {
    "registry": "https://registry.npmjs.org/",
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/journeyapps-labs/micro.git"
  },
  "author": "JOURNEYAPPS",
  "license": "Apache-2.0",
  "scripts": {
    "test": "vitest"
  },
  "files": [
    "dist/**"
  ],
  "dependencies": {
    "@apidevtools/json-schema-ref-parser": "^14.2.1",
    "@journeyapps-labs/micro-codecs": "workspace:^",
    "@journeyapps-labs/micro-errors": "workspace:^",
    "ajv": "^8.17.1",
    "better-ajv-errors": "^2.0.2",
    "ts-codec": "^1.3.0",
    "vitest": "^3.2.4",
    "zod": "^4.1.12"
  }
}
