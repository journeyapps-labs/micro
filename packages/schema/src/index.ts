export * from './json-schema/keywords';
export * from './json-schema/parser';

export * from './definitions';
export * from './validators';
export * from './utils';
