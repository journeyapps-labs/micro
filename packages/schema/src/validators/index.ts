export * from './ts-codec-validator';
export * from './schema-validator';
export * from './zod-validator';
