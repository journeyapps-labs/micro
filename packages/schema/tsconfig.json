{
  "extends": "../../tsconfig.base",
  "compilerOptions": {
    "outDir": "dist",
    "declarationDir": "dist",
    "rootDir": "src",
    "resolveJsonModule": true
  },
  "include": [
    "./src/**/*.ts",
    "./more/schema/**/*.json"
  ],
  "references": [
    {
      "path": "../codecs"
    },
    {
      "path": "../errors"
    }
  ]
}
