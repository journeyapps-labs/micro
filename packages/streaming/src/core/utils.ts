import * as micro_schema from "@journeyapps-labs/micro-schema";
import * as micro_errors from "@journeyapps-labs/micro-errors";
import * as cross_stream from "./cross-stream";
import type * as webstreams from "stream/web";

export type StreamLike<T> = Iterable<T> | AsyncIterable<T>;

/**
 * Construct an AsyncIterator from a given ReadableStream.
 *
 * This is only really intended to be used from browser runtimes or within code intended to
 * be used cross-labs. This is because Node ReadableStreams already implement AsyncIterators
 */
export async function* iterableFromReadable<T = any>(
  readable: ReadableStream<T> | webstreams.ReadableStream<T>,
) {
  const reader = readable.getReader();

  try {
    while (true) {
      const res = await reader.read();
      if (res.done) {
        return;
      }
      yield res.value;
    }
  } finally {
    reader.releaseLock();
  }
}

/**
 * Construct a ReadableStream from a given Iterable or AsyncIterable.
 *
 * If the given iterable is already a readable then this is a noop
 */
export const readableFrom = <T = any>(
  iterable: StreamLike<T>,
  strategy?: QueuingStrategy<T | undefined>,
): webstreams.ReadableStream<T> => {
  if (iterable instanceof cross_stream.Readable) {
    return iterable;
  }

  let resume: (() => void) | undefined;
  return new cross_stream.Readable<T>(
    {
      start(controller) {
        void (async function () {
          for await (const chunk of iterable) {
            controller.enqueue(chunk);

            if (controller.desiredSize != null && controller.desiredSize <= 0) {
              await new Promise<void>((resolve) => {
                resume = resolve;
              });
            }
          }

          controller.close();
        })().catch((err) => {
          controller.error(err);
        });
      },
      async pull() {
        resume?.();
      },
    },
    strategy,
  );
};

/**
 * Yield a generator that validates data flowing through it
 */
export async function* validateDataStream<T>(
  iterable: StreamLike<T>,
  validator: micro_schema.MicroValidator<T>,
) {
  for await (const chunk of iterable) {
    const res = validator.validate(chunk);
    if (!res.valid) {
      throw new micro_errors.ValidationError(res.errors);
    }
    yield chunk;
  }
}
