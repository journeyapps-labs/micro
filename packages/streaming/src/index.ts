export * as compat from './core/cross-stream';
export * as middleware from './middleware';
export * as bson from './bson';
export * from './core';
