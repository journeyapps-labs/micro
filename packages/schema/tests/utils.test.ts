import { describe, test, expect } from 'vitest';

import * as utils from '../src/utils';
import schema from './fixtures/schema';

describe('utils', () => {
  test('allow additional properties in a json schema', () => {
    const cleaned_schema = utils.allowAdditionalProperties(schema);
    expect(cleaned_schema).toMatchSnapshot();
  });

  test('it should only modify additionalProperties if it is a boolean', () => {
    const cleaned_schema = utils.allowAdditionalProperties({
      definitions: {
        a: {
          type: 'object',
          properties: {
            prop: {
              type: 'string'
            }
          },
          additionalProperties: false,
          required: ['prop']
        }
      },

      type: 'object',
      properties: {
        name: {
          type: 'object',
          additionalProperties: {
            $ref: '#/definitions/a'
          }
        }
      },
      required: ['name', 'b'],
      additionalProperties: false
    });
    expect(cleaned_schema).toMatchSnapshot();
  });
});
