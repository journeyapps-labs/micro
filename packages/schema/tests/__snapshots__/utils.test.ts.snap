// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`utils > allow additional properties in a json schema 1`] = `
{
  "definitions": {
    "c": {
      "properties": {
        "prop": {
          "type": "string",
        },
      },
      "required": [
        "prop",
      ],
      "type": "object",
    },
  },
  "properties": {
    "b": {
      "oneOf": [
        {
          "properties": {
            "a": {
              "type": "number",
            },
          },
          "required": [
            "a",
          ],
          "type": "object",
        },
      ],
    },
    "d": {
      "$ref": "#/definitions/c",
    },
    "name": {
      "properties": {
        "a": {
          "type": "string",
        },
        "b": {
          "enum": [
            "A",
          ],
        },
      },
      "required": [
        "a",
      ],
      "type": "object",
    },
  },
  "required": [
    "name",
    "b",
  ],
  "type": "object",
}
`;

exports[`utils > it should only modify additionalProperties if it is a boolean 1`] = `
{
  "definitions": {
    "a": {
      "properties": {
        "prop": {
          "type": "string",
        },
      },
      "required": [
        "prop",
      ],
      "type": "object",
    },
  },
  "properties": {
    "name": {
      "additionalProperties": {
        "$ref": "#/definitions/a",
      },
      "type": "object",
    },
  },
  "required": [
    "name",
    "b",
  ],
  "type": "object",
}
`;
